R/caretStack.R

```diff
@@ -55,80 +54,95 @@ caretStack <- function(all.models, ...){
+#'   methodList = c("rpart", "glm")
 #' )
-#' meta_model <- caretStack(models, method="lm")
-#' RMSE(predict(meta_model, iris[101:150,1:2]), iris[101:150,3])
+#' meta_model <- caretStack(models, method = "lm")
+#' RMSE(predict(meta_model, iris[101:150, 1:2]), iris[101:150, 3])
 #' }
 predict.caretStack <- function(
-  object, newdata=NULL,
-  se=FALSE, level=0.95,
-  return_weights=FALSE,
-  na.action=na.omit,
-  ...){
+    object, newdata = NULL,
+    se = FALSE, level = 0.95,
+    return_weights = FALSE,
+    na.action = na.omit,
+    ...) {
   stopifnot(is(object$models, "caretList"))
   type <- extractModelTypes(object$models)
 
-  #preds <- do.call(predict, c(list(object=object$models, newdata=newdata), modelList_predict_params)) TODO keep this around for future use
-  preds <- predict(object$models, newdata=newdata, na.action=na.action)
-  if(type == "Classification"){
-    out <- predict(object$ens_model, newdata=preds, na.action=na.action, ...)
-    # Need a check here
-    if(class(out) %in% c("data.frame", "matrix")){
-      # Return probability predictions for only one of the classes
-      # as determined by configured default response class level
-      est <- out[, getBinaryTargetLevel(), drop = TRUE]
-    } else{
-      est <- out
+  # preds <- do.call(predict, c(list(object=object$models, newdata=newdata), modelList_predict_params)) TODO keep this around for future use
+  preds <- predict(object$models, newdata = newdata, na.action = na.action)
+
+  if (type == "Classification") {
+    column_names <- colnames(preds)
+    num_classes <- length(levels(object$models[[1]]$pred$obs))
+    if (getMulticlassExcludedLevel() >= 1 && getMulticlassExcludedLevel() <= num_classes) {
+      classes_included <- levels(object$models[[1]]$pred$obs)[-getMulticlassExcludedLevel()]
+    } else {
+      classes_included <- levels(object$models[[1]]$pred$obs)
     }
-  } else{
-    est <- predict(object$ens_model, newdata=preds, ...)
+    pattern <- paste(classes_included, collapse = "|")
+    # Remove columns that are associated with the class that was excluded
+    filtered_column_names <- grep(pattern, column_names, value = TRUE)
+    preds <- preds[, filtered_column_names, drop = FALSE]
+
+    est <- predict(object$ens_model, newdata = preds, na.action = na.action, ...)
+  } else {
+    est <- predict(object$ens_model, newdata = preds, ...)
   }
 
-  if(se | return_weights){
+  if (se || return_weights) {
     imp <- varImp(object$ens_model)$importance
-    weights <- imp$Overall
-    weights[!is.finite(weights)] <- 0
-    weights <- weights / sum(weights)
-
-    names(weights) <- row.names(imp)
+    weights <- as.list(as.data.frame(imp))
     methods <- colnames(preds)
-
-    for(m in setdiff(methods, names(weights))){
-      weights[m] <- 0
-    }
+    weights <- lapply(weights, function(class_weights) {
+      # ensure that we have a numeric vector
+      ifelse(is.finite(class_weights), class_weights, 0)
```

Comment:
this line seems odd to me: `ifelse(is.finite(class_weights), class_weights, 0)`

Should it be something like `class_weights <- ifelse(is.finite(class_weights), class_weights, 0)`

Reply:
Yes, you are right.